include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,501 @@
+                                          Strategy const& strategy)
+    {
+        for (auto pit1 = points_begin(polygon); pit1 != points_end(polygon); ++pit1)
+        {
+            for (auto pit2 = pit1; pit2 != points_end(polygon); ++pit2)
+            {
+                for (auto pit3 = pit2; pit3 != points_end(polygon); ++pit3)
+                {
+                    if (affinely_independent<0, 1>(*pit1, *pit2, *pit3, strategy)
+                        || affinely_independent<0, 2>(*pit1, *pit2, *pit3, strategy)
+                        || affinely_independent<1, 2>(*pit1, *pit2, *pit3, strategy))
+                    {
+                        set_vec_from_point::apply(*pit1, base);
+                        set_vec_from_point::apply(*pit2, v1);
+                        set_vec_from_point::apply(*pit3, v2);
+                        return true;
+                    }
+                }
+            }
+        }
+        return false; // All points are collinear
+    }
+
+    static point_2d project_point_to_plane_2d(vec3 const& v1, vec3 const& v2, vec3 const& p)
+    {
+        return point_2d{qvm::dot(v1, p), qvm::dot(v2, p)};
+    }
+
+    template <typename VisitPolicy>
+    static bool has_correct_intersection_orientation(polygon_2d const& polygon,
+                                                     point_2d const& point1,
+                                                     point_2d const& point2,
+                                                     VisitPolicy& visitor)
+    {
+        auto ring = boost::geometry::exterior_ring(polygon);
+        if (boost::geometry::closure<ring_2d>::value == boost::geometry::closed)
+        {
+            ring.pop_back(); // Remove the closing point to make it open
+        }
+
+        for (auto it = boost::begin(ring); it != boost::end(ring); ++it)
+        {
+            if (boost::geometry::equals(*it, point1))
+            {
+                if (boost::geometry::equals(point1, point2))
+                {
+                    visitor.template apply<failure_duplicate_points>();
+                    return false;
+                }
+                auto prev = (it == boost::begin(ring)) ? boost::end(ring) - 1 : boost::prior(it);
+                if (boost::geometry::equals(point2, *prev))
+                {
+                    // valid intersection and consistent orientation
+                    return true;
+                }
+                auto next = (it == boost::end(ring)) ? boost::begin(ring) : boost::next(it);
+                if (boost::geometry::equals(point2, *next))
+                {
+                    // valid intersection but inconsistent orientation
+                    visitor.template apply<failure_inconsistent_orientation>();
+                    return false;
+                }
+            }
+        }
+        visitor.template apply<failure_invalid_intersection>();
+        return false;
+    }
+
+    // Check if a segment intersects a polygon in 3D space
```

Comment:
So also here `point1` and `point2` form a segment, it would be good to make that clear.